ibis/backends/exasol/tests/conftest.py

```diff
@@ -0,0 +1,118 @@
+from __future__ import annotations
+
+import os
+import subprocess
+from typing import TYPE_CHECKING
+
+import ibis
+from ibis.backends.tests.base import RoundHalfToEven, ServiceBackendTest, UnorderedComparator
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+    from pathlib import Path
+    from typing import Any
+
+EXASOL_USER = os.environ.get("IBIS_TEST_EXASOL_USER", "sys")
+EXASOL_PASS = os.environ.get("IBIS_TEST_EXASOL_PASSWORD", "exasol")
+EXASOL_HOST = os.environ.get("IBIS_TEST_EXASOL_HOST", "localhost")
+EXASOL_PORT = int(os.environ.get("IBIS_TEST_EXASOL_PORT", 8563))
+IBIS_TEST_EXASOL_DB = os.environ.get("IBIS_TEST_EXASOL_DATABASE", "EXASOL")
+
+
+class TestConf(UnorderedComparator, ServiceBackendTest, RoundHalfToEven):
```

Comment:
Hi @cpcloud,

While digging into test issues with the Exasol backend, I stumbled upon a few things that got me curious. I was wondering if you may, could shed some light on them.

- **Stable Order in Default Expectations**: Why is the default expectation for a stable order present if sorting isn't explicitly requested (`ORDER BY`)? Is this because most backends tend to provide stable results?
- **Different Ways of Customizing Backend Tests**: What is the rationale behind having two different approaches to customizing backend tests (i.e., using inheritance and using class member-based settings)?

  Why not only use the class members approach? E.g.: 
  ```python
  TestConf(UnorderedComparator, ServiceBackendTest):
       ...
  ```
  VS
  ```python
   TestConf(ServiceBackendTest):
       comparator = UnorderedComparator
  ```
- **Limited Values in string_col of functional_alltypes**: Why does the `string_col` in `functional_alltypes` only contain values from `[0..9]`? This limitation doesn't seem beneficial for testing functions like `CAPITALIZE` or `LOWER`, and it also lacks a wide range of edge cases. Could this be expanded for more comprehensive testing?

Just trying to get a better understanding of the workspace because I feel like I've only grasped a small part of it so far.

Thanks a lot for all the help, I really appreciate it!

best
Nico

Reply:
Hey @Nicoretti -- I can answer a few of these:

> * **Different Ways of Customizing Backend Tests**: What is the rationale behind having two different approaches to customizing backend tests (i.e., using inheritance and using class member-based settings)?

I don't believe there's a rationale here, they're artifacts of the long evolution of the project.  Further, many (most? all?) of the test class mixins can probably be merged into the base `BackendTest` class.  



> * **Limited Values in string_col of functional_alltypes**: Why does the `string_col` in `functional_alltypes` only contain values from `[0..9]`? This limitation doesn't seem beneficial for testing functions like `CAPITALIZE` or `LOWER`, and it also lacks a wide range of edge cases. Could this be expanded for more comprehensive testing?

It should _absolutely_ be expanded for more comprehensive testing.

Reply:
To answer the first question:

> Stable Order in Default Expectations

Many backends do provide a stable ordering, but some don't and you can override the `assert_series_equal` and `assert_frame_equal` methods on `BackendTest` as you like. The BigQuery backend is an example of a backend that doesn't have a stable ordering without `ORDER BY` and overrides those methods appropriately.